src/psum.c

```diff
@@ -0,0 +1,184 @@
+#include "data.table.h"
+
+// like base::p{max,min}, but for sum
+SEXP psum(SEXP x, SEXP narmArg) {
+  if (!isNewList(x)) error(_("Internal error: x must be a list"));
+  if (!isLogical(narmArg) || LENGTH(narmArg)!=1 || LOGICAL(narmArg)[0]==NA_LOGICAL) error(_("na.rm must be TRUE or FALSE"));
+
+  int J=LENGTH(x);
+  SEXPTYPE outtype = INTSXP;
+  int n = -1;
+  for (int j=0; j<J; j++) {
+    SEXP xj=VECTOR_ELT(x, j);
+    switch(TYPEOF(xj)) {
+    case INTSXP:
+      break;
+    case REALSXP:
+      outtype = REALSXP;
+      break;
+    default:
+      error(_("Only numeric inputs are supported for psum"));
+    }
+    if (n >= 0) {
+      if (n != LENGTH(xj)) {
+        error(_("Inconsistent input lengths -- first found %d, but %d element has length %d"), n, j+1, LENGTH(xj));
+      }
+    } else { // initialize
+      n = LENGTH(xj);
+    }
+  }
+
+
+  // TODO: support int->double conversion
```

Comment:
I think it make sense to **not** provide automatic conversion, until we will have #3765

Reply:
OK, done the following for now:

1. when `na.rm=FALSE`, overflowed elements are `NA`, with warning.
1. when `na.rm=TRUE`, error if overflow happens. I suppose we could make the output here `NA` as well, but it would require some sort of refactor -- at the moment, in the `na.rm=TRUE` case, `NA` is used to seed the output as a shortcut to get the `psum(NA, NA, na.rm=TRUE) == NA` behavior working. If there's more than one route to becoming `NA` in this case, it'll need some logic adjustment to account for that.